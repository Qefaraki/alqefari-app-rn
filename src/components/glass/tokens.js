// Design tokens for the liquid-glass aesthetic
// Centralized so we can tune the look consistently across components

export const glassTokens = {
  radii: {
    xs: 12,
    sm: 16,
    md: 20,
    lg: 28,
    xl: 36,
  },
  blurIntensity: {
    subtle: 28,
    medium: 40,
    strong: 64,
  },
  // Base surface overlay colors for a bright UI
  surfaceOverlay: {
    // Vertical overlay from slightly brighter top to slightly dimmer bottom
    gradient: ['rgba(255,255,255,0.55)', 'rgba(255,255,255,0.28)'],
  },
  // Hairline borders that feel like condensed water on the glass
  borders: {
    inner: 'rgba(255,255,255,0.55)',
    outer: 'rgba(0,0,0,0.04)',
  },
  // Soft ambient elevation â€“ avoid harsh shadows
  elevation: {
    ios: {
      shadowColor: '#000',
      shadowOpacity: 0.10,
      shadowRadius: 14,
      shadowOffset: { width: 0, height: 8 },
    },
    android: {
      elevation: 3,
    },
  },
  highlight: {
    // Specular highlight sweeping from top-left
    sheen: ['rgba(255,255,255,0.7)', 'rgba(255,255,255,0.0)'],
  },
};

export default glassTokens;


