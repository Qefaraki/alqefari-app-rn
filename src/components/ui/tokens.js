// Design tokens for neo-native (non-glass) UI
// Neutral surfaces, soft elevation, crisp dividers, Arabic-first
// Aligned with iOS Human Interface Guidelines

const tokens = {
  colors: {
    bg: '#F2F2F7',
    surface: '#FFFFFF',
    text: '#111827',
    textMuted: '#6B7280',
    divider: 'rgba(0,0,0,0.08)',
    outline: 'rgba(0,0,0,0.12)',
    accent: '#007AFF',
    success: '#34C759',
    danger: '#FF3B30',
    najdi: {
      background: '#F9F7F3', // Al-Jass White
      container: '#D1BBA3', // Camel Hair Beige
      text: '#242121', // Sadu Night
      textMuted: '#736372',
      primary: '#A13333', // Najdi Crimson
      secondary: '#D58C4A', // Desert Ochre
      focus: '#957EB5',
    },
    diff: {
      added: '#34C759',       // iOS green for new values
      addedBg: '#34C75915',   // 8% opacity
      removed: '#FF3B30',     // iOS red for old values
      removedBg: '#FF3B3015', // 8% opacity
      modified: '#007AFF',    // iOS blue for changes
      modifiedBg: '#007AFF15', // 8% opacity
    },
  },
  // iOS-standard typography scale
  typography: {
    largeTitle: { fontSize: 34, fontWeight: '700', lineHeight: 41 },
    title1: { fontSize: 28, fontWeight: '700', lineHeight: 34 },
    title2: { fontSize: 22, fontWeight: '700', lineHeight: 28 },
    title3: { fontSize: 20, fontWeight: '600', lineHeight: 25 },
    headline: { fontSize: 17, fontWeight: '600', lineHeight: 22 },
    body: { fontSize: 17, fontWeight: '400', lineHeight: 22 },
    callout: { fontSize: 16, fontWeight: '400', lineHeight: 21 },
    subheadline: { fontSize: 15, fontWeight: '400', lineHeight: 20 },
    footnote: { fontSize: 13, fontWeight: '400', lineHeight: 18 },
    caption1: { fontSize: 12, fontWeight: '400', lineHeight: 16 },
    caption2: { fontSize: 11, fontWeight: '400', lineHeight: 13 },
  },
  radii: {
    sm: 10,
    md: 12,
    lg: 16,
    xl: 20,
    full: 9999,  // Full pill shape for chips
  },
  // iOS-standard spacing scale (8px grid)
  spacing: {
    xxs: 4,
    xs: 8,
    sm: 12,
    md: 16,
    lg: 20,
    xl: 24,
    xxl: 32,
  },
  // iOS minimum touch target
  touchTarget: {
    minimum: 44,
  },
  shadow: {
    ios: {
      shadowColor: '#000',
      shadowOpacity: 0.07,
      shadowRadius: 12,
      shadowOffset: { width: 0, height: 8 },
    },
    android: {
      elevation: 3,
    },
  },
};

export default tokens;
