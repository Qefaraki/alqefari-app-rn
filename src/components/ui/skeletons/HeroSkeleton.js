import React from 'react';
import { View, StyleSheet } from 'react-native';
import Shimmer from './Shimmer';
import tokens from '../tokens';

/**
 * HeroSkeleton - Loading skeleton for ProfileViewer Hero section
 * Matches Hero.js structure with photo, name, lineage, and metrics
 */
const HeroSkeleton = ({ withPhoto = true }) => {
  return (
    <View
      style={styles.container}
      accessible={true}
      accessibilityLabel="جاري تحميل الملف الشخصي"
      accessibilityRole="progressbar"
    >
      {/* Hero Photo */}
      {withPhoto && (
        <View style={styles.photoWrapper}>
          <Shimmer
            width="100%"
            height={220}
            borderRadius={0}
          />
        </View>
      )}

      <View style={styles.body}>
        {/* Name */}
        <View style={styles.nameRow}>
          <Shimmer
            width="60%"
            height={32}
            borderRadius={tokens.radii.sm}
          />
          {!withPhoto && (
            <View style={styles.actionsPlaceholder}>
              <Shimmer width={40} height={40} borderRadius={20} />
            </View>
          )}
        </View>

        {/* Lineage / Common Name */}
        <View style={styles.chainRow}>
          <Shimmer
            width="75%"
            height={18}
            borderRadius={tokens.radii.sm}
          />
        </View>

        {/* Bio lines (optional) */}
        <View style={styles.bioContainer}>
          <Shimmer
            width="100%"
            height={16}
            borderRadius={tokens.radii.sm}
            style={styles.bioLine}
          />
          <Shimmer
            width="90%"
            height={16}
            borderRadius={tokens.radii.sm}
            style={styles.bioLine}
          />
          <Shimmer
            width="70%"
            height={16}
            borderRadius={tokens.radii.sm}
          />
        </View>

        {/* Metrics Row */}
        <View style={styles.metricsRow}>
          <Shimmer
            width={100}
            height={64}
            borderRadius={tokens.radii.md}
            style={styles.metricPill}
          />
          <Shimmer
            width={100}
            height={64}
            borderRadius={tokens.radii.md}
            style={styles.metricPill}
          />
          <Shimmer
            width={100}
            height={64}
            borderRadius={tokens.radii.md}
          />
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'relative',
    borderRadius: 20,
    overflow: 'hidden',
    marginBottom: 24,
    backgroundColor: tokens.colors.najdi.background, // #F9F7F3
  },
  photoWrapper: {
    height: 220,
    backgroundColor: tokens.colors.najdi.container + '30',
  },
  body: {
    paddingHorizontal: 20,
    paddingVertical: 20,
  },
  nameRow: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginBottom: 12,
  },
  actionsPlaceholder: {
    marginLeft: 12,
  },
  chainRow: {
    marginBottom: 12,
  },
  bioContainer: {
    marginTop: 12,
    marginBottom: 16,
  },
  bioLine: {
    marginBottom: 8,
  },
  metricsRow: {
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 16,
    gap: 10,
  },
  metricPill: {
    flex: 1,
    maxWidth: 120,
  },
});

export default HeroSkeleton;
