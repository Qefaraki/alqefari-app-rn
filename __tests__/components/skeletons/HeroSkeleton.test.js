/**
 * HeroSkeleton Component Test Suite
 *
 * Tests the hero section loading skeleton
 */

import React from 'react';
import { render } from '@testing-library/react-native';
import HeroSkeleton from '../../../src/components/ui/skeletons/HeroSkeleton';
import tokens from '../../../src/components/ui/tokens';

// Mock dependencies
jest.mock('../../../src/components/ui/skeletons/Shimmer', () => 'Shimmer');

describe('HeroSkeleton Component', () => {
  describe('Rendering', () => {
    it('renders without crashing', () => {
      const { toJSON } = render(<HeroSkeleton />);
      expect(toJSON()).toBeTruthy();
    });

    it('renders with photo by default', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // With photo: photo shimmer + name + lineage + 3 bio lines + 3 metrics = 9 shimmers
      expect(shimmers.length).toBeGreaterThanOrEqual(7);
    });

    it('renders without photo when withPhoto=false', () => {
      const { UNSAFE_root } = render(<HeroSkeleton withPhoto={false} />);

      // Find photo wrapper - should not exist
      const photoWrappers = UNSAFE_root.findAllByProps({ style: expect.objectContaining({ height: 220 }) });

      // Should not find photo wrapper with exact height of 220
      const hasPhotoShimmer = photoWrappers.some(wrapper => {
        try {
          wrapper.findByType('Shimmer');
          return true;
        } catch {
          return false;
        }
      });

      expect(hasPhotoShimmer).toBe(false);
    });

    it('renders action button placeholder when withPhoto=false', () => {
      const { UNSAFE_root } = render(<HeroSkeleton withPhoto={false} />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Find circular shimmer (action button) - 40x40 with borderRadius 20
      const actionButton = shimmers.find(s =>
        s.props.width === 40 &&
        s.props.height === 40 &&
        s.props.borderRadius === 20
      );

      expect(actionButton).toBeTruthy();
    });
  });

  describe('Structure and Layout', () => {
    it('renders name row with correct shimmer dimensions', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Name shimmer: 60% width, 32 height
      const nameShimmer = shimmers.find(s =>
        s.props.width === '60%' && s.props.height === 32
      );

      expect(nameShimmer).toBeTruthy();
    });

    it('renders lineage row with correct dimensions', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Lineage shimmer: 75% width, 18 height
      const lineageShimmer = shimmers.find(s =>
        s.props.width === '75%' && s.props.height === 18
      );

      expect(lineageShimmer).toBeTruthy();
    });

    it('renders bio lines with varying widths', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Bio lines: 100%, 90%, 70% width, all 16 height
      const bioLine1 = shimmers.find(s =>
        s.props.width === '100%' && s.props.height === 16
      );
      const bioLine2 = shimmers.find(s =>
        s.props.width === '90%' && s.props.height === 16
      );
      const bioLine3 = shimmers.find(s =>
        s.props.width === '70%' && s.props.height === 16
      );

      expect(bioLine1).toBeTruthy();
      expect(bioLine2).toBeTruthy();
      expect(bioLine3).toBeTruthy();
    });

    it('renders three metric pills', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Metric pills: 100 width, 64 height, md borderRadius
      const metricPills = shimmers.filter(s =>
        s.props.width === 100 &&
        s.props.height === 64 &&
        s.props.borderRadius === tokens.radii.md
      );

      expect(metricPills).toHaveLength(3);
    });

    it('renders photo shimmer with correct dimensions', () => {
      const { UNSAFE_root } = render(<HeroSkeleton withPhoto={true} />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Photo shimmer: 100% width, 220 height, 0 borderRadius
      const photoShimmer = shimmers.find(s =>
        s.props.width === '100%' &&
        s.props.height === 220 &&
        s.props.borderRadius === 0
      );

      expect(photoShimmer).toBeTruthy();
    });
  });

  describe('Design System Compliance', () => {
    it('uses Najdi background color', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const container = UNSAFE_root.findByProps({ style: expect.objectContaining({ backgroundColor: tokens.colors.najdi.background }) });

      expect(container).toBeTruthy();
    });

    it('uses correct border radius (20)', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const container = UNSAFE_root.findByProps({ style: expect.objectContaining({ borderRadius: 20 }) });

      expect(container).toBeTruthy();
    });

    it('uses tokens.radii.sm for small shimmers', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Name and lineage use sm radius
      const nameShimmer = shimmers.find(s =>
        s.props.width === '60%' && s.props.borderRadius === tokens.radii.sm
      );
      const lineageShimmer = shimmers.find(s =>
        s.props.width === '75%' && s.props.borderRadius === tokens.radii.sm
      );

      expect(nameShimmer).toBeTruthy();
      expect(lineageShimmer).toBeTruthy();
    });

    it('uses tokens.radii.md for metric pills', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      const metricPills = shimmers.filter(s =>
        s.props.borderRadius === tokens.radii.md
      );

      expect(metricPills.length).toBeGreaterThanOrEqual(3);
    });
  });

  describe('Responsive Behavior', () => {
    it('maintains structure with photo', () => {
      const { toJSON } = render(<HeroSkeleton withPhoto={true} />);
      expect(toJSON()).toMatchSnapshot();
    });

    it('maintains structure without photo', () => {
      const { toJSON } = render(<HeroSkeleton withPhoto={false} />);
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('Edge Cases', () => {
    it('handles withPhoto=undefined (defaults to true)', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Should include photo shimmer by default
      const photoShimmer = shimmers.find(s =>
        s.props.height === 220
      );

      expect(photoShimmer).toBeTruthy();
    });

    it('handles withPhoto=null', () => {
      const { toJSON } = render(<HeroSkeleton withPhoto={null} />);
      expect(toJSON()).toBeTruthy();
    });

    it('handles withPhoto=true explicitly', () => {
      const { UNSAFE_root } = render(<HeroSkeleton withPhoto={true} />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      const photoShimmer = shimmers.find(s => s.props.height === 220);
      expect(photoShimmer).toBeTruthy();
    });
  });

  describe('Performance', () => {
    it('renders within performance budget (< 16ms)', () => {
      const start = performance.now();
      render(<HeroSkeleton />);
      const end = performance.now();

      const renderTime = end - start;
      expect(renderTime).toBeLessThan(16);
    });

    it('renders efficiently with photo', () => {
      const start = performance.now();
      render(<HeroSkeleton withPhoto={true} />);
      const end = performance.now();

      expect(end - start).toBeLessThan(20);
    });

    it('renders efficiently without photo', () => {
      const start = performance.now();
      render(<HeroSkeleton withPhoto={false} />);
      const end = performance.now();

      expect(end - start).toBeLessThan(20);
    });
  });

  describe('Component Count', () => {
    it('renders correct number of shimmers with photo', () => {
      const { UNSAFE_root } = render(<HeroSkeleton withPhoto={true} />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Photo(1) + Name(1) + Lineage(1) + Bio(3) + Metrics(3) = 9
      expect(shimmers).toHaveLength(9);
    });

    it('renders correct number of shimmers without photo', () => {
      const { UNSAFE_root } = render(<HeroSkeleton withPhoto={false} />);
      const shimmers = UNSAFE_root.findAllByType('Shimmer');

      // Name(1) + Action(1) + Lineage(1) + Bio(3) + Metrics(3) = 9
      expect(shimmers).toHaveLength(9);
    });
  });

  describe('Accessibility', () => {
    it('renders with proper container structure', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const container = UNSAFE_root.findByProps({ style: expect.objectContaining({ position: 'relative' }) });

      expect(container).toBeTruthy();
    });

    it('includes overflow:hidden on container', () => {
      const { UNSAFE_root } = render(<HeroSkeleton />);
      const container = UNSAFE_root.findByProps({ style: expect.objectContaining({ overflow: 'hidden' }) });

      expect(container).toBeTruthy();
    });
  });
});
