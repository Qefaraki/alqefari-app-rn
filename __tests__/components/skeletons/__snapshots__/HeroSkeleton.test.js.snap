// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HeroSkeleton Component Responsive Behavior maintains structure with photo 1`] = `
<View
  style={
    {
      "backgroundColor": "#F9F7F3",
      "borderRadius": 20,
      "marginBottom": 24,
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#D1BBA330",
        "height": 220,
      }
    }
  >
    <Shimmer
      borderRadius={0}
      height={220}
      width="100%"
    />
  </View>
  <View
    style={
      {
        "paddingHorizontal": 20,
        "paddingVertical": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 12,
        }
      }
    >
      <Shimmer
        borderRadius={10}
        height={32}
        width="60%"
      />
    </View>
    <View
      style={
        {
          "marginBottom": 12,
        }
      }
    >
      <Shimmer
        borderRadius={10}
        height={18}
        width="75%"
      />
    </View>
    <View
      style={
        {
          "marginBottom": 16,
          "marginTop": 12,
        }
      }
    >
      <Shimmer
        borderRadius={10}
        height={16}
        style={
          {
            "marginBottom": 8,
          }
        }
        width="100%"
      />
      <Shimmer
        borderRadius={10}
        height={16}
        style={
          {
            "marginBottom": 8,
          }
        }
        width="90%"
      />
      <Shimmer
        borderRadius={10}
        height={16}
        width="70%"
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 10,
          "marginTop": 16,
        }
      }
    >
      <Shimmer
        borderRadius={12}
        height={64}
        style={
          {
            "flex": 1,
            "maxWidth": 120,
          }
        }
        width={100}
      />
      <Shimmer
        borderRadius={12}
        height={64}
        style={
          {
            "flex": 1,
            "maxWidth": 120,
          }
        }
        width={100}
      />
      <Shimmer
        borderRadius={12}
        height={64}
        width={100}
      />
    </View>
  </View>
</View>
`;

exports[`HeroSkeleton Component Responsive Behavior maintains structure without photo 1`] = `
<View
  style={
    {
      "backgroundColor": "#F9F7F3",
      "borderRadius": 20,
      "marginBottom": 24,
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <View
    style={
      {
        "paddingHorizontal": 20,
        "paddingVertical": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 12,
        }
      }
    >
      <Shimmer
        borderRadius={10}
        height={32}
        width="60%"
      />
      <View
        style={
          {
            "marginLeft": 12,
          }
        }
      >
        <Shimmer
          borderRadius={20}
          height={40}
          width={40}
        />
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 12,
        }
      }
    >
      <Shimmer
        borderRadius={10}
        height={18}
        width="75%"
      />
    </View>
    <View
      style={
        {
          "marginBottom": 16,
          "marginTop": 12,
        }
      }
    >
      <Shimmer
        borderRadius={10}
        height={16}
        style={
          {
            "marginBottom": 8,
          }
        }
        width="100%"
      />
      <Shimmer
        borderRadius={10}
        height={16}
        style={
          {
            "marginBottom": 8,
          }
        }
        width="90%"
      />
      <Shimmer
        borderRadius={10}
        height={16}
        width="70%"
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 10,
          "marginTop": 16,
        }
      }
    >
      <Shimmer
        borderRadius={12}
        height={64}
        style={
          {
            "flex": 1,
            "maxWidth": 120,
          }
        }
        width={100}
      />
      <Shimmer
        borderRadius={12}
        height={64}
        style={
          {
            "flex": 1,
            "maxWidth": 120,
          }
        }
        width={100}
      />
      <Shimmer
        borderRadius={12}
        height={64}
        width={100}
      />
    </View>
  </View>
</View>
`;
